### PR TITLE
s3: optimize info/exists calls

### PR DESCRIPTION
**Pull request by isidentical**
Described in https://github.com/dask/s3fs/issues/470#issuecomment-829097599

### CODE REVIEW

**Comment by martindurant** [comment]
Please ping me to have a thorough review of this.

First a couple of thoughts:
- loosing version information could be a dealbraker for some, so we would only want to do this when the filesystem is not marked as version-aware
- for the case where the user does not have list permission, we would still want to fall back to HEAD (but not bother trying to list for the directory)

**Comment by isidentical** [comment]
> (but not bother trying to list for the directory)

AFAIK permissions might allow searching for prefixes that ends with trailing slash, so it kinda makes sense to bother to list directory since that time we'll call it with a trailing slash.

**Comment by martindurant** [comment]
There is a separate permission for listing files with a prefix versus listing with a prefix that happens to end with "/"? That would be annoying!

**Comment by martindurant** [comment]
This might be a good place to include a test against https://github.com/dask/s3fs/issues/473 and fix, if necessary.

**Comment by isidentical** [comment]
> There is a separate permission for listing files with a prefix versus listing with a prefix that happens to end with "/"?

I think we hit that before, though @efiop probably has more context than me. It is not a specific permission per se, but rather a bucket-level rule declaration that allows you to list `/bucket/something/*` but not `/bucket/somethingXXX` etc.

**Comment by isidentical** [comment]
> This might be a good place to include a test against #473 and fix, if necessary.

That seems a very different bug, though since it is trivial a fix is included in this PR.

**Comment by efiop** [comment]
> > There is a separate permission for listing files with a prefix versus listing with a prefix that happens to end with "/"?
> 
> I think we hit that before, though @efiop probably has more context than me. It is not a specific permission per se, but rather a bucket-level rule declaration that allows you to list `/bucket/something/*` but not `/bucket/somethingXXX` etc.

Correct, here is a ticket for it https://github.com/iterative/dvc/issues/5130

**Comment by martindurant** [comment]
Can you please also add the test shown in https://github.com/dask/s3fs/issues/475 here? I suspect the changes in this PR will have fixed that.

**Comment by isidentical** [comment]
> Can you please also add the test shown in #475 here? I suspect the changes in this PR will have fixed that.

If I am not missing anything, I already did https://github.com/dask/s3fs/pull/472/commits/0eb3620b788775c1a4badec6f2861dad5e53a748

File: s3fs/core.py [hunk 1]
```diff
@@ -888,6 +889,62 @@ async def _get_file(self, rpath, lpath, version_id=None):
...
+                "LastModified": file["LastModified"],
+                "Size": file["Size"],
+                "size": file["Size"],
+                "name": "/".join([bucket, key]),
+                "type": "file",
+                "StorageClass": "STANDARD",
+                "VersionId": file.get("VersionId"),
+            }
+        elif not (file or out.get("KeyCount", 0) or out.get("CommonPrefixes", [])):
+            raise FileNotFoundError(path)
```
Comment (martindurant) [inline s3fs/core.py:919]: Aren't there `else` cases here:
- we got a file by a different name, so the original file was not found
- we got prefix(es), so the path is a directory (does MaxKeys restrict the number of prefixes? Why use delimiter in the call at all?)
- we got a path which is the same except for an extra "/"

What happens if we fail all the ifs, and return `None`?

**Comment by martindurant** [comment]
Specifically the test
```python
@pytest.mark.parametrize("find", [False, True])
def test_lsdir(s3, find):
    if find:
        s3.find(test_bucket_name)

    d = test_bucket_name + "/test"
    assert d in s3.ls(test_bucket_name)
```

**Comment by isidentical** [comment]
> This might be a good place to include a test against #473 and fix, if necessary.
> Can you please also add the test shown in #475 here? I suspect the changes in this PR will have fixed that.

wait you are talking about different issues?

**Comment by isidentical** [comment]
this PR doesn't interact with ls/find in any way, so I don't think there is any reason to include a fix for #475.

**Comment by martindurant** [comment]
OK, that will have to be dealt with separately.

File: s3fs/core.py (hunk 1 above)
Comment (isidentical) [reply s3fs/core.py:919]: > What happens if we fail all the ifs, and return None?

The `else: ` is the condition where the prefix doesn't exist as is, and there are other prefixes. For example for a directory like this;
```
very/
   a12/
      another
   a1234
```

`very/a123` will be handled in the first condition, `very/bsomething` will be handled in the second condition since no such prefix exists either as a file or a directory and `very/a1` and `very/a2` will fallback to the directory search which would return true for `very/a2` and raise file not found error for `very/a`.

I couldn't find a test case which isn't handled by this logic

Comment (isidentical) [reply s3fs/core.py:919]: No, else is a totally different path. For example there might other prefixes matched, so it might be a directory etc

File: s3fs/core.py [hunk 2]
```diff
@@ -908,27 +965,17 @@ async def _info(self, path, bucket=None, key=None, refresh=False, version_id=Non
...
+                    out = await self._version_aware_info(path, version_id)
+                else:
+                    try:
+                        out = await self._simple_info(path)
+                    except PermissionError:
+                        # If the permissions aren't enough for scanning a prefix
+                        # then fall back to using normal HEAD_OBJECT
+                        out = await self._version_aware_info(path, version_id)
+                if out:
+                    return out
```
Comment (martindurant) [inline s3fs/core.py:978]: So, is `out` ever `None` or empty here, or would that have raised by now? This is the end of the function anyway, so the `if` seems redundant. It would be nice to specify what this method can be expected to return - indeed, I should have done this.

**Review by martindurant — COMMENTED** [review]
> No, else is a totally different path.

You mean there are three branches here: the return statement above (key found), this branch (we know the file doesn't exist) and a third implicit branch where we return `None`?

File: s3fs/core.py (hunk 2 above)
Comment (isidentical) [reply s3fs/core.py:978]: It can be None indeed, if it fall backs to the third (implicit) branch.

Comment (martindurant) [reply s3fs/core.py:978]: OK, got it. Thanks for clarifying - for some reason I found it hard to follow that. That all sounds good, maybe you want to add a comment in the code - or not.

**Comment by isidentical** [comment]
Okay, I'll add a few comments

**Comment by martindurant** [comment]
(I'll merge later - going out prospecting for a vaccine)

**Comment by isidentical** [comment]
it should be fine

**Comment by martindurant** [comment]
Done! Thanks for this.
